File: Data Structures & Algorithms/target-sum/submission-3.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        dp = {}
        n = len(nums)
        def dfs(i, currsum):
            if(i >= n):
                if(currsum == target):
                    return 1
                else:
                    return 0
            if (i,currsum) in dp:
                return dp[(i,currsum)]
            
            dp[(i,currsum)]= dfs(i+1, currsum+nums[i])+dfs(i+1,currsum-nums[i])
            return dp[(i,currsum)]
        return dfs(0,0)
        

File: Data Structures & Algorithms/find-first-and-last-position-of-element-in-sorted-array/submission-3.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int l = 0;
        int r = nums.size();
        int mid;
        if(nums.size() == 0){
            return vector<int> {-1,-1};
        }
        while(l<r){
            mid = l + (r-l)/2;
            if(nums[mid] == target){
                int start = l;
                int end = r;
                while(nums[start]!=target){
                    start++;
                }
                while(nums[end]!=target){
                    end--;
                }
                return vector<int> {start,end};
            }
            else if(nums[mid] > target){
                r = mid;
            }
            else{
                l = mid+1;
            }
        }
        return vector<int> {-1,-1};
    }
};